database/src/database.rs

```diff
@@ -149,6 +149,41 @@ pub trait DatabaseRw<T: Table>: DatabaseRo<T> {
     /// This will return [`RuntimeError::KeyNotFound`] wrapped in [`Err`] if `key` does not exist.
     fn delete(&mut self, key: &T::Key) -> Result<(), RuntimeError>;
 
+    /// Delete and return a key-value pair in the database.
+    ///
+    /// This is the same as [`DatabaseRw::delete`], however,
+    /// it will serialize the `T::Value` and return it.
+    ///
+    /// # Errors
+    /// This will return [`RuntimeError::KeyNotFound`] wrapped in [`Err`] if `key` does not exist.
+    fn take(&mut self, key: &T::Key) -> Result<T::Value, RuntimeError>;
+
+    /// Fetch the value, and apply a function to it - or delete the entry.
+    ///
+    /// This will call [`DatabaseRo::get`] and call your provided function `f` on it.
+    ///
+    /// - If `f` returns `Some(value)`, that will be [`DatabaseRw::put`] as the new value
+    /// - If `f` returns `None`, the entry will be [`DatabaseRw::delete`]d
+    ///
+    /// # Errors
+    /// This will return [`RuntimeError::KeyNotFound`] wrapped in [`Err`] if `key` does not exist.
+    fn update<F>(&mut self, key: &T::Key, mut f: F) -> Result<(), RuntimeError>
+    where
+        F: FnMut(T::Value) -> Option<T::Value>,
+    {
+        let value = DatabaseRo::get(self, key)?;
+
+        match f(value) {
+            Some(value) => DatabaseRw::put(self, key, &value)?,
+            None => match DatabaseRw::delete(self, key) {
+                Ok(()) | Err(RuntimeError::KeyNotFound) => (),
+                Err(e) => return Err(e),
```

Comment:
Took 30 sec to wonder what would be the use case for `None` resulting items to be deleted. Can you add to the doc a little mention that it can be used for conditional deletion ?

Reply:
And I also don't see how we can fall into `RuntimeError::KeyNotFound` since we just `get`d it earlier.

Reply:
> Can you add to the doc a little mention that it can be used for conditional deletion ?

Added. I think this function could be replaced by some `std::collections::btree_map::Entry`-like API but that can be for later.

>  And I also don't see how we can fall into RuntimeError::KeyNotFound since we just getd it earlier.

True, removed.